### PR TITLE
Generate random outgoing fragment for share tracking

### PR DESCRIPTION
**Pull request by muxin**
#3135

### CODE REVIEW

File: src/utils/base64.js [hunk 1]
```diff
@@ -61,11 +66,49 @@ export function base64Decode(str) {
...
+ * @param {string} str
+ * @return {string}
+ */
+export function base64UrlEncode(str) {
+  return btoa(str).replace(/[+/=]/g, ch => base64UrlEncodeSubs[ch]);
+}
+
+/**
+ * Converts a bytes array into base64url encoded string.
+* base64url is defined in RFC 4648. It is sometimes referred to as "web safe".
```
Comment (lannka) [inline src/utils/base64.js]: missing space

File: src/utils/bytes.js [hunk 2]
```diff
@@ -32,3 +32,27 @@ export function stringToBytes(str) {
...
+ * Generate a random bytes array with specific length
+ * @param {!number} length
+ * @return {!Uint8Array}
+ */
+export function getRandomBytesArray(length) {
+  let random = Math.random();
+  const bytes = new Uint8Array(length);
+  for (let i = 0; i < length; i++) {
+    random *= 256;
+    bytes[i] = Math.floor(random);
```
Comment (lannka) [inline src/utils/bytes.js]: you need
```
random -= bytes[i];
```
otherwise it will overflow

File: test/functional/utils/test-base64.js [hunk 3]
```diff
@@ -74,3 +76,27 @@ describe('base64DecodeToBytes', () => {
     expect(() => base64DecodeToBytes('c3Vy=')).to.throw();
   });
 });
+
+describe('base64EncodeFromBytes', () => {
+  it('should encode a bytes array to base64url string correctly', () => {
+    expect(base64UrlEncodeFromBytes(new Uint8Array()))
+      .to.equal('');
+    expect(base64UrlEncodeFromBytes(stringToBytes('s')))
+      .to.equal('cw..');
```
Comment (lannka) [inline test/functional/utils/test-base64.js]: try to find examples with those special chars

File: test/functional/utils/test-base64.js [hunk 4]
```diff
@@ -74,3 +76,27 @@ describe('base64DecodeToBytes', () => {
...
+      .to.equal('c3U.');
+    expect(base64UrlEncodeFromBytes(stringToBytes('sur')))
+      .to.equal('c3Vy');
+  });
+
+  it('should encode a bytes array to base64 string correctly', () => {
+    expect(base64EncodeFromBytes(new Uint8Array()))
+      .to.equal('');
+    expect(base64EncodeFromBytes(stringToBytes('s')))
+      .to.equal('cw==');
```
Comment (lannka) [inline test/functional/utils/test-base64.js]: ditto

File: extensions/amp-share-tracking/0.1/test/test-amp-share-tracking.js [hunk 5]
```diff
@@ -31,6 +31,7 @@ describe('amp-share-tracking', () => {
     sandbox = sinon.sandbox.create();
     viewerForMock = sandbox.stub(Viewer.prototype, 'getFragment');
     xhrMock = sandbox.stub(Xhr.prototype, 'fetchJson');
+    sandbox.stub(Math, 'random', () => 0.111111111111111111111111111111);
```
Comment (lannka) [inline extensions/amp-share-tracking/0.1/test/test-amp-share-tracking.js]: nit: you can put some random numbers at the end, to make it clear that we truncate the string

File: test/functional/utils/test-bytes.js [hunk 6]
```diff
@@ -33,4 +46,22 @@ describe('stringToBytes', function() {
     }).to.throw();
   });
 
+  it('should convert bytes array to string', () => {
+    const string = bytesToString(new Uint8Array([102, 111, 111]));
```
Comment (lannka) [inline test/functional/utils/test-bytes.js]: rename `string` to `str`. `string` is a reserved key word

File: test/functional/utils/test-bytes.js [hunk 7]
```diff
@@ -33,4 +46,22 @@ describe('stringToBytes', function() {
...
+    expect(string).to.equal('foo');
+  });
+
+  it('should generate random bytes array', () => {
+    sandbox.stub(Math, 'random').returns(0.111111111111111111111111111111);
```
Comment (lannka) [inline test/functional/utils/test-bytes.js]: no need to use sandbox.


`Math.random = () => 0.11111111111111111111111;` does the same job

File: src/utils/bytes.js (hunk 2 above)
Comment (muxin) [inline src/utils/bytes.js]: Thanks for pointing that out~ fixed

File: test/functional/utils/test-base64.js (hunk 3 above)
Comment (muxin) [inline test/functional/utils/test-base64.js]: added another example with special chars

File: test/functional/utils/test-base64.js (hunk 4 above)
Comment (muxin) [inline test/functional/utils/test-base64.js]: done

File: extensions/amp-share-tracking/0.1/test/test-amp-share-tracking.js (hunk 5 above)
Comment (muxin) [inline extensions/amp-share-tracking/0.1/test/test-amp-share-tracking.js]: ok, done

File: test/functional/utils/test-bytes.js (hunk 6 above)
Comment (muxin) [inline test/functional/utils/test-bytes.js]: done

File: test/functional/utils/test-bytes.js (hunk 7 above)
Comment (muxin) [inline test/functional/utils/test-bytes.js]: done

**Comment by lannka** [comment]
LGTM

File: test/functional/utils/test-bytes.js [hunk 8]
```diff
@@ -16,7 +16,10 @@
 
 import {
   stringToBytes,
+  bytesToString,
+  getRandomBytesArray,
 } from '../../../src/utils/bytes';
+import * as sinon from 'sinon';
```
Comment (lannka) [inline test/functional/utils/test-bytes.js]: not needed now

Comment (muxin) [inline test/functional/utils/test-bytes.js]: done

File: src/utils/bytes.js [hunk 9]
```diff
@@ -32,3 +32,28 @@ export function stringToBytes(str) {
...
+export function bytesToString(bytes) {
+  return String.fromCharCode.apply(String, bytes);
+};
+
+/**
+ * Generate a random bytes array with specific length
+ * @param {!number} length
+ * @return {!Uint8Array}
+ */
+export function getRandomBytesArray(length) {
```
Comment (jridgewell) [inline src/utils/bytes.js]: This is me a very random bytes array, and should be named to it's not used with anything that does crypto.

File: test/functional/utils/test-bytes.js [hunk 10]
```diff
@@ -33,4 +35,22 @@ describe('stringToBytes', function() {
     }).to.throw();
   });
 
+  it('should convert bytes array to string', () => {
+    const str = bytesToString(new Uint8Array([102, 111, 111]));
+    expect(str).to.equal('foo');
+  });
+
+  it('should generate random bytes array', () => {
+    Math.random = () => 0.1111111111111111;
```
Comment (jridgewell) [inline test/functional/utils/test-bytes.js]: You need to stub, everything that executes afterwards is now using a borked `Math.random` now!

Comment (lannka) [inline test/functional/utils/test-bytes.js]: We can just move this to `beforeEach` to make it clear. I think we can avoid the use of sinon here.

File: src/utils/bytes.js (hunk 9 above)
Comment (lannka) [inline src/utils/bytes.js]: `getPseudoRandomBytesArray` ?

Comment (jridgewell) [inline src/utils/bytes.js]: Psuedo random still means random. This is a predictable pattern, not random at all.

File: test/functional/utils/test-bytes.js (hunk 10 above)
Comment (jridgewell) [inline test/functional/utils/test-bytes.js]: Whichever.

File: src/utils/bytes.js (hunk 9 above)
Comment (lannka) [inline src/utils/bytes.js]: > The Math.random() function returns a floating-point, pseudo-random number in the range (0, 1) 

This is a quote from [MDN doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random) for `Math.random()`.

I think "pseudo" implies "predictable". Or you have better name?

Comment (jridgewell) [inline src/utils/bytes.js]: What I mean is the start number is random, then the pattern it generates is predictable. Contrast that with `Crypto#getRandomValues`, which has the same method signature but has no pattern (it's also a PRNG). Using `Math.random` `n` times would be better and match `Crypto`'s implementation (just without the crypto random guarantee).

Comment (lannka) [inline src/utils/bytes.js]: You are totally correct. But here we are using `Math.random()` only once.

Actually, the thing we can really criticize here is that the function doesn't support big `length`.

Comment (muxin) [inline src/utils/bytes.js]: @jridgewell What do you mean by using `Math.random` n times? Is `n` the length of the bytes array? Each time, multiply the result of Math.random with 256 and use `Math.floor` to generate one random byte?

File: test/functional/utils/test-bytes.js (hunk 10 above)
Comment (muxin) [inline test/functional/utils/test-bytes.js]: I could use stub here. Will make change in another PR.

File: src/utils/bytes.js (hunk 9 above)
Comment (jridgewell) [inline src/utils/bytes.js]:
```js
export function getRandomBytesArray(length) {
  var bytes = new Uint8Array(length);
  for (var i = 0; i < length; i++) {
    bytes[i] = Math.random() * 256;
  }
  return bytes;
}

Comment (muxin) [inline src/utils/bytes.js]: OK, cool. The only concern I have is: whether it's time-consuming to use `Math.random` n times? In the share-tracking case, we are only using length of 6, which I guess is fine. @lannka Do you think this is fine?

Comment (lannka) [inline src/utils/bytes.js]: hmm, what i learned from security folks is that all consequent `Math.random` calls are predictable, which means no extra entropy. So it doesn't really give you more bytes. 

That's to say, when you compare two such arrays that were generated in this way, if the first byte of the arrays equals, all the remaining equal to each other as well.

Comment (muxin) [inline src/utils/bytes.js]: So from what I understand, we cannot use `Math.random` n times within a single call because that will lower the entropy (first byte determines the rest bytes, only 256 possible outputs)

Comment (jridgewell) [inline src/utils/bytes.js]: > hmm, what i learned from security folks is that all consequent Math.random calls are predictable, which means no extra entropy. So it doesn't really give you more bytes.

That's the "psuedo" part. But it's still much more random than what we're doing here. https://medium.com/@betable/tifu-by-using-math-random-f1c308c4fd9d#.zi48jhb65

Comment (muxin) [inline src/utils/bytes.js]: After discussion with @jridgewell and @lannka, we agreed to use the win.crypto.getRandomValues (which has enough entropy) by default if available, otherwise, fall back to use Math.random once only in the share tracking code(`Math.random` has limited precision and it's enough for random share tracking generation). The corresponding PR is #4678
